README.md
The nodes included in this release are in development for the X3DOM environment to support
mobile devices and VR environments. The intent is that these be included in X3D V4

Documentation for V4 nodes is at http://tools.realism.com/specification/x3d-v40
Documentation for specific nodes is at http://tools.realism.com/specification/x3d-v40/changes-additions-x3d-v33


Welcome to X3DOM 
================
X3DOM is a framework for integrating and manipulating (X)3D scenes as HTML5
DOM elements, which are rendered via an X3D plugin, Flash/Stage3D or WebGL.
The open-source system allows defining 3D scene description and runtime
behavior declaratively, without any low-level JavaScript or GLSL coding.

This addition is a testbed platform for X3D V4 nodes. It is developed as
an addition to X3DOM 

This library should be included after all standard X3DOM libraries.

What follows below is the documentation for X3DOM V1.7.1

Getting started
---------------
Please have a look at the official documentation in order to get started:

http://doc.x3dom.org


Need help?
----------
If you want to use X3DOM and seek help and exchange of knowledge, please use
the x3dom-users mailing list or the forums. Beginner questions usually have
been asked already and in many cases the mailinglist or forum archives provide
quick answers to many questions.

If you decide to post to the mailing list or forum, please include as much
info as possible and describe your problem and what you tried already
succinctly.


Mailing List
-------------
Please use the mailinglists as much as you can. The 
[x3dom-users][] list is the one you are most likely interested in 
and the first address for help or questions.

Questions specifically about V4 nodes and features should be directed
to the x3d-public@web3d.org mailing list.


Issue Tracker
-------------
Please report issues and attach patches here. For general help and questions
direct your inquiry to the x3dom-users mailing list or one of the forums.

https://github.com/x3dom/x3dom/issues


**Uploading Files to the Issue Tracker**   
The GitHub issue tracker does not allow to upload files. However, there is a
service that helps you with sharing files: https://gist.github.com. Just copy
and paste your file contents there and then copy the Gist URL into the
issues form.

It immensely helps us if you can provide a live web page illustrating your
problem. So if you have webspace and can upload and post the URL with your
issue, the chances we will have a look at it increase.


Contributing
------------
We encourage you to contribute to X3DOM! If you would like to contribute to
X3DOM, you should subscribe to the developers mailing list ([x3dom-dev][]), 
where the developers of X3DOM discuss the development of the library itself.

You can send pull requests via GitHub. Patches should:

  1. Follow the style of the existing code.
  2. One commit should do exactly one thing.
  3. Commit messages should start with a summary line below 80 characters 
     followed by a blank line, and then the reasoning/analysis for why the 
     change was made (if appropriate).
  4. Commits that fix a bug in a previous commit (which has already been 
     merged) should start with `fixup!` and then the summary line of the 
     commit it fixes.
  5. Rebase your branch against the upstream’s master. We don’t want to pull 
     redundant merge commits.
  6. **Be clear about what license applies to your patch:** The files within 
     this repository are under the GPL3 (or later) but, as the original 
     creators, we are still allowed to create non-free derivatives. However, 
     if patches are given to us under GPL then those cannot make it into any 
     non-free derivatives we may later wish to create. So to make it easier 
     for us (and avoid any legal issues) we prefer if patches are released as 
     public domain.


### GitHub Workflow

Developing patches should follow this workflow:

  1.  Fork on GitHub (click Fork button)
  2.  Clone to computer: `git clone git@github.com:«github account»/x3dom.git`
  3.  cd into your repo: `cd x3dom`
  4.  Set up remote upstream: `git remote add -f upstream git://github.com/x3dom/x3dom.git`
  5.  Create a branch for the new feature: `git checkout -b my_new_feature`
  6.  Work on your feature, add and commit as usual

Creating a branch is not strictly necessary, but it makes it easy to delete 
your branch when the feature has been merged into upstream, diff your branch 
with the version that actually ended in upstream, and to submit pull requests 
for multiple features (branches).

  7.  Push branch to GitHub: `git push origin my_new_feature`
  8.  Issue pull request: Click Pull Request button on GitHub

#### Useful Commands

If a lot of changes have happened upstream you can replay your local changes 
on top of these, this is done with `rebase`, e.g.:

    git fetch upstream
    git rebase upstream/master

This will fetch changes and re-apply your commits on top of these.

This is generally better than merge, as it will give a clear picture of which 
commits are local to your branch. It will also “prune” any of your local 
commits if the same changes have been applied upstream.

You can use `-i` with `rebase` for an “interactive” rebase. This allows you 
to drop, re-arrange, merge, and reword commits, e.g.:

	git rebase -i upstream/master

There's always the [x3dom-dev][] mailinglist.


[x3dom-users]:   https://lists.sourceforge.net/lists/listinfo/x3dom-users
[x3dom-dev]:     https://lists.sourceforge.net/lists/listinfo/x3dom-developers
[GPL 3]:         http://www.gnu.org/copyleft/gpl.html


RELEASENOTES.rst
.. _releasenotes:

Release notes
=============

Version 0.02
-------------

* Alpha releaase of Macro node
* Only 'url' and 'params' fields are implemented

Version 0.01
-------------

* Initial release
* Created new nodes Animate and Interpolate. Nodes still being developed.


test/lib/browserFlags.js
/*
 * browserFlags.js
 *
 *	Processes parameters and flags that control the display
 *
 */
var counterShader = 0;
var urlSearch = window.location.search.substring(1,window.location.search.length);
var nvPairs = urlSearch.split('&');
var tf = {'1':true, 'true':true, 'TRUE':true, '0':false, 'false':false, 'FALSE':false};
var flags = {
			DEBUG :		0,
			DESKTOP :	0,
			VIEW :		'Viewer',
			X3D :		1
			};
for (ii=0; ii<nvPairs.length; ii++) {
	nv = nvPairs[ii].split('=');
	flags[nv[0].toUpperCase()] = nv[1];
}
flags.showStats	= tf[flags.DEBUG];
flags.showLog	= tf[flags.DEBUG];
flags.DESKTOP	= tf[flags.DESKTOP];
flags.X3D		= tf[flags.X3D];

if (flags.DESKTOP) {
	//alert ('Loading jQuery-ui');
	document.writeln ('<scri' + 'pt src="lib/jquery-ui-1.11.4/jquery-ui.js"></scr' + 'ipt>');
	document.writeln ('<link rel="stylesheet" type="text/css" href="lib/jquery-ui-1.11.4/jquery-ui.css">');
}


src/nodes/Time/Animate.js
/** @namespace x3dom.nodeTypes */
/*
 * X3DOM JavaScript Library
 * http://www.x3dom.org
 *
 * (C)2009 Fraunhofer IGD, Darmstadt, Germany
 * Dual licensed under the MIT and GPL
 */

// ### Animate ###
x3dom.registerNodeType(
    "Animate",
    "Time",
    defineClass(x3dom.nodeTypes.X3DSensorNode,
        
        /**
         * Constructor for Animate
         * @constructs x3dom.nodeTypes.Animate
         * @x3d 3.3
         * @component Time
         * @status full
         * @extends x3dom.nodeTypes.X3DSensorNode
         * @param {Object} [ctx=null] - context object, containing initial settings like namespace
         * @classdesc Animate nodes generate events as time passes.
         */
        function (ctx) {
            x3dom.nodeTypes.Animate.superClass.call(this, ctx);

            if (ctx)
                ctx.doc._nodeBag.timer.push(this);
            else
                x3dom.debug.logWarning("Animate: No runtime context found!");


            /**
             * The "cycle" of a Animate node lasts for cycleInterval seconds. The value of cycleInterval shall be greater than zero.
             * @var {x3dom.fields.SFTime} cycleInterval
             * @range [0, inf]
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 1
             * @field x3d
             * @instance
             */
            this.addField_SFTime(ctx, 'cycleInterval', 1);


            /**
             * Specifies whether the timer cycle loops.
             * @var {x3dom.fields.SFBool} loop
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue false
             * @field x3d
             * @instance
             */
            this.addField_SFBool(ctx, 'loop', false);

            /**
             * Sets the startTime for the cycle.
             * @var {x3dom.fields.SFTime} startTime
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
             */
            this.addField_SFTime(ctx, 'startTime', 0);

            /**
             * Sets a time for the timer to stop.
             * @var {x3dom.fields.SFTime} stopTime
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
             */
            this.addField_SFTime(ctx, 'stopTime', 0);

            /**
             * Sets a time for the timer to pause.
             * @var {x3dom.fields.SFTime} pauseTime
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
             */
            this.addField_SFTime(ctx, 'pauseTime', 0);

            /**
             * Sets a time for the timer to resume from pause.
             * @var {x3dom.fields.SFTime} resumeTime
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
             */
            this.addField_SFTime(ctx, 'resumeTime', 0);


            /**
             * A cycleTime outputOnly field can be used for synchronization purposes such as sound with animation.
             * The value of a cycleTime event will be equal to the time at the beginning of the current cycle. A cycleTime event is generated at the beginning of every cycle, including the cycle starting at startTime.
             * The first cycleTime event for a Animate node can be used as an alarm (single pulse at a specified time).
             * @var {x3dom.fields.SFTime} cycleTime
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
			 *
			 *	Not a field (event) for Animation. Leftover from TimeSensor.
             */
            //this.addField_SFTime(ctx, 'cycleTime', 0);

            /**
             * The elapsedTime outputOnly field delivers the current elapsed time since the Animate was activated and running, cumulative in seconds and not counting any time while in a paused state.
             * @var {x3dom.fields.SFTime} elapsedTime
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
			 *
			 *	Not a field (event) for Animation. Leftover from TimeSensor.
             */
            //this.addField_SFTime(ctx, 'elapsedTime', 0);

            /**
             * fraction_changed events output a floating point value in the closed interval [0, 1]. At startTime the value of fraction_changed is 0. After startTime, the value of fraction_changed in any cycle will progress through the range (0.0, 1.0].
             * @var {x3dom.fields.SFFloat} fraction_changed
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
			 *
			 *	Not a field (event) for Animation. Leftover from TimeSensor.
             */
            //this.addField_SFFloat(ctx, 'fraction_changed', 0);

            /**
             * Outputs whether the timer is active.
             * @var {x3dom.fields.SFBool} isActive
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue false
             * @field x3d
             * @instance
             */
            this.addField_SFBool(ctx, 'isActive', false);

            /**
             * Outputs whether the timer is paused.
             * @var {x3dom.fields.SFBool} isPaused
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue false
             * @field x3d
             * @instance
             */
            this.addField_SFBool(ctx, 'isPaused', false);

            /**
             * The time event sends the absolute time for a given tick of the Animate node.
             * @var {x3dom.fields.SFTime} time
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0
             * @field x3d
             * @instance
			 *
			 *	Field (event) not implemented for Animation. May not exist in final form.
             */
            //this.addField_SFTime(ctx, 'time', 0);

            /**
             * Starts Animate -- inputOnly
             * @var {x3dom.fields.SFBool} start
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue true
             * @field x3dv4
             * @instance
             */
            this.addField_SFBool(ctx, 'start', true);

            /**
             * Stop Animate -- inputOnly
             * @var {x3dom.fields.SFBool} stop
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue false
             * @field x3dv4
             * @instance
             */
            this.addField_SFBool(ctx, 'stop', false);

            /**
             * Animation maximum frame rate (frames per second)
             * @var {x3dom.fields.SFBool} maxFrameRate
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 30
             * @field x3dv4
             * @instance
             */
            this.addField_SFFloat(ctx, 'maxFrameRate', 30);

            /**
             * Interpolator nodes
             * @var {x3dom.fields.SFBool} interpolate
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue []
             * @field x3dv4
             * @instance
             */
            this.addField_MFNode('interpolate', x3dom.nodeTypes.X3DInterpolatorNode);

            /**
             *
             * @var {x3dom.fields.SFBool} first
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue true
             * @field x3dom
             * @instance
             */
            this.addField_SFBool(ctx,'first', true);

            /**
             *
             * @var {x3dom.fields.SFFloat} firstCycle
             * @memberof x3dom.nodeTypes.Animate
             * @initvalue 0.0
             * @field x3dom
             * @instance
             */
            this.addField_SFFloat(ctx,'firstCycle', 0.0);

            this._prevCycle = -1;
            this._lastTime = 0;
            this._cycleStopTime = 0;
            this._activatedTime = 0;

            if (this._vf.startTime > 0) {
                this._updateCycleStopTime();
            }

            this._backupStartTime = this._vf.startTime;
            this._backupStopTime = this._vf.stopTime;
            this._backupCycleInterval = this._vf.cycleInterval;
			this._frameInterval = 1.0/Math.max(1.0, Math.min(1000.0, this._vf.maxFrameRate));
        
        },
        {
            tick: function (time)
            {
                if (!this._vf.enabled) {
                    this._lastTime = time;
                    return false;
                }
				
				delta = time-this._lastRenderedTime;
//				console.log ('Delta = ' + delta + '; min Interval = ' + this._frameInterval + '\n');
				if (delta < this._frameInterval) {
					return false;
				}
//				console.log ('Animate at time ' + time+'\n');
				this._lastRenderedTime = time;

                var isActive = ( this._vf.cycleInterval > 0 &&
                    time >= this._vf.startTime &&
                    (time < this._vf.stopTime || this._vf.stopTime <= this._vf.startTime) &&
                    (this._vf.loop == true || (this._vf.loop == false && time < this._cycleStopTime)) );

                if (isActive && !this._vf.isActive) {
                    this.postMessage('isActive', true);
                    this._activatedTime = time;
                }

                // Checking for this._vf.isActive allows the dispatch of 'final events' (before deactivation)
                if (isActive || this._vf.isActive) {
                    //this.postMessage('elapsedTime', time - this._activatedTime);

                    var isPaused = ( time >= this._vf.pauseTime && this._vf.pauseTime > this._vf.resumeTime );

                    if (isPaused && !this._vf.isPaused) {
                        this.postMessage('isPaused', true);
                        //this.postMessage('pauseTime', time);
                    } else if (!isPaused && this._vf.isPaused) {
                        this.postMessage('isPaused', false);
                        //this.postMessage('resumeTime', time);
                    }

                    if (!isPaused) {
                        var cycleFrac = this._getCycleAt(time);
                        var cycle = Math.floor(cycleFrac);

                        var cycleTime = this._vf.startTime + cycle*this._vf.cycleInterval;
                        var adjustTime = 0;

                        if (this._vf.stopTime > this._vf.startTime &&
                            this._lastTime < this._vf.stopTime && time >= this._vf.stopTime)
                            adjustTime = this._vf.stopTime;
                        else if (this._lastTime < cycleTime && time >= cycleTime)
                            adjustTime = cycleTime;

                        if( adjustTime > 0 ) {
                            time = adjustTime;
                            cycleFrac = this._getCycleAt(time);
                            cycle = Math.floor(cycleFrac);
                        }

                        var fraction = cycleFrac - cycle;
						//console.log ('Animate at time ' + time + ' -- ' + fraction + '\n');
						
						for (ii=0; ii<this._cf.interpolate.nodes.length; ii++) {
							this._cf.interpolate.nodes[ii].doInterpolate(fraction);
						}

                        if (fraction < x3dom.fields.Eps) {
                            fraction = ( this._lastTime < this._vf.startTime ? 0.0 : 1.0 );
                            //this.postMessage('cycleTime', time);
                        }

                        //this.postMessage('fraction_changed', fraction);
                        //this.postMessage('time', time);
                    }
                }

                if (!isActive && this._vf.isActive)
                    this.postMessage('isActive', false);

                this._lastTime = time;

                return true;
            },

            fieldChanged: function(fieldName)
            {
                if (fieldName == "enabled") {
                    // TODO; eval other relevant outputs
                    if (!this._vf.enabled && this._vf.isActive) {
                        this.postMessage('isActive', false);
                    }
                }
                else if (fieldName == "startTime") {
                    // Spec: Should be ignored when active. (Restore old value)
                    if (this._vf.isActive) {
                        this._vf.startTime = this._backupStartTime;
                        return;
                    }

                    this._backupStartTime = this._vf.startTime;
                    this._updateCycleStopTime();
                }
                else if (fieldName == "stopTime") {
                    // Spec: Should be ignored when active and less than startTime. (Restore old value)
                    if (this._vf.isActive && this._vf.stopTime <= this._vf.startTime) {
                        this._vf.stopTime = this._backupStopTime;
                        return;
                    }

                    this._backupStopTime = this._vf.stopTime;
                }
                else if (fieldName == "cycleInterval") {
                    // Spec: Should be ignored when active. (Restore old value)
                    if (this._vf.isActive) {
                        this._vf.cycleInterval = this._backupCycleInterval;
                        return;
                    }

                    this._backupCycleInterval = this._vf.cycleInterval;
                }
                else if (fieldName == "loop") {
                    this._updateCycleStopTime();
                }
            },

            parentRemoved: function(parent)
            {
                if (this._parentNodes.length === 0) {
                    var doc = this.findX3DDoc();

                    for (var i=0, n=doc._nodeBag.timer.length; i<n; i++) {
                        if (doc._nodeBag.timer[i] === this) {
                            doc._nodeBag.timer.splice(i, 1);
                        }
                    }
                }
            },

            _getCycleAt: function(time)
            {
                return Math.max( 0.0, time - this._vf.startTime ) / this._vf.cycleInterval;
            },

            _updateCycleStopTime: function()
            {
                if (this._vf.loop == false) {
                    var now = new Date().getTime() / 1000;
                    var cycleToStop = Math.floor(this._getCycleAt(now)) + 1;

                    this._cycleStopTime = this._vf.startTime + cycleToStop*this._vf.cycleInterval;
                }
                else {
                    this._cycleStopTime = 0;
                }
            }
        }
    )
);

src/nodes/Networking/Macro.js
/** @namespace x3dom.nodeTypes */
/*
 * X3DOM JavaScript Library
 * http://www.x3dom.org
 *
 * (C)2009 Fraunhofer IGD, Darmstadt, Germany
 * Dual licensed under the MIT and GPL
 */

// ### Macro ###
x3dom.registerNodeType(
    "Macro",
    "Networking",
    defineClass(x3dom.nodeTypes.X3DGroupingNode,
        
        /**
         * Constructor for Macro
		 * Definition: http://tools.realism.com/specification/x3d-v40/abstract-specification/changes-additions-x3d-v33/macro
         * @constructs x3dom.nodeTypes.Macro
         * @x3d 3.3
         * @component Networking
         * @status full
         * @extends x3dom.nodeTypes.X3DGroupingNode
         * @param {Object} [ctx=null] - context object, containing initial settings like namespace
         * @classdesc Macro is a Grouping node that can load nodes from another X3D scene via url.
         */
        function (ctx) {
            x3dom.nodeTypes.Macro.superClass.call(this, ctx);


            /**
             * Each specified URL shall refer to a valid X3D file that contains a list of children nodes, prototypes and routes at the top level. Hint: Strings can have multiple values, so separate each string by quote marks. Warning: strictly match directory and filename capitalization for http links!
             * @var {x3dom.fields.MFString} url
             * @memberof x3dom.nodeTypes.Macro
             * @initvalue []
             * @field x3d
             * @instance
             */
            this.addField_MFString(ctx, 'url', []);

            /**
             * The X3D node string to insert into the scene graph. This must be a valid X3D fragment that contains a list of children nodes and routes at the top level. 
             * @var {x3dom.fields.SFString} expand
             * @memberof x3dom.nodeTypes.Macro
             * @initvalue ""
             * @field x3d
             * @instance
             */
            this.addField_MFString(ctx, 'expand', "");

            /**
             * Specifies whether the X3D file specified by the url field is loaded. Hint: use LoadSensor to detect when loading is complete. TRUE: load immediately (it's also possible to load the URL at a later time by sending a TRUE event to the load field); FALSE: no action is taken (by sending a FALSE event to the load field of a previously loaded Macro, the contents of the Macro will be unloaded from the scene graph)
             * @var {x3dom.fields.SFBool} load
             * @memberof x3dom.nodeTypes.Macro
             * @initvalue true
             * @field x3d
             * @instance
             */
            this.addField_SFBool(ctx, 'load', true);

            /**
             * Specifies the namespace of the Macro node.
             * @var {x3dom.fields.MFString} nameSpaceName
             * @memberof x3dom.nodeTypes.Macro
             * @initvalue []
             * @field x3dom
             * @instance
             */
            this.addField_MFString(ctx, 'nameSpaceName', ['_internal']);

            /**
             * Specifies the substituion parameters of the Macro node.
             * @var {x3dom.fields.MFString} params
             * @memberof x3dom.nodeTypes.Macro
             * @initvalue []
             * @field x3dom
             * @instance
             */
            this.addField_MFString(ctx, 'params', []);

            /**
             * Specifies whether the DEF value is used as id when no other id is set.
             * @var {x3dom.fields.SFBool} mapDEFToID
             * @memberof x3dom.nodeTypes.Macro
             * @initvalue false
             * @field x3dom
             * @instance
             */
            this.addField_SFBool(ctx, 'mapDEFToID', false);

            this.initDone = false;
            this.count = 0;
            this.numRetries = x3dom.nodeTypes.Macro.MaximumRetries;
        
        },
        {
            fieldChanged: function (fieldName)
            {
                if (fieldName == "url") {

                    //Remove the childs of the x3domNode
                    for (var i=0; i<this._childNodes.length; i++)
                    {
                        this.removeChild(this._childNodes[i]);
                    }

                    //if reflected to DOM remove the childs of the domNode
                    if (this._vf.nameSpaceName.length != 0) {
                        var node = this._xmlNode;
                        if (node && node.hasChildNodes())
                        {
                            while ( node.childNodes.length >= 1 )
                            {
                                node.removeChild( node.firstChild );
                            }
                        }
                    }
                    this.loadMacro();
                }
                else if (fieldName == "render") {
                    this.invalidateVolume();
                    //this.invalidateCache();
                }
            },

            nodeChanged: function ()
            {
                if (!this.initDone) {
                    this.initDone = true;
                    this.loadMacro();
                }
            },

            fireEvents: function(eventType)
            {
                if ( this._xmlNode &&
                    (this._xmlNode['on'+eventType] ||
                        this._xmlNode.hasAttribute('on'+eventType) ||
                        this._listeners[eventType]) )
                {
                    var event = {
                        target: this._xmlNode,
                        type: eventType,
                        error: (eventType == "error") ? "XMLHttpRequest Error" : "",
                        cancelBubble: false,
                        stopPropagation: function() { this.cancelBubble = true; }
                    };

                    try {
                        var attrib = this._xmlNode["on" + eventType];

                        if (typeof(attrib) === "function") {
                            attrib.call(this._xmlNode, event);
                        }
                        else {
                            var funcStr = this._xmlNode.getAttribute("on" + eventType);
                            var func = new Function('event', funcStr);
                            func.call(this._xmlNode, event);
                        }

                        var list = this._listeners[eventType];
                        if (list) {
                            for (var i = 0; i < list.length; i++) {
                                list[i].call(this._xmlNode, event);
                            }
                        }
                    }
                    catch(ex) {
                        x3dom.debug.logException(ex);
                    }
                }
            },

/*
 * Loads a external Xml resource into a string
 */
            loadExternalXml: function ()
            {
                var that = this;

                var xhr = new window.XMLHttpRequest();
                if (xhr.overrideMimeType)
                    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
				//xhr.resultSuccess = success;

                xhr.onreadystatechange = function ()
                {
                    if (xhr.readyState != 4) {
                        // still loading
                        //x3dom.debug.logInfo('Loading inlined data... (readyState: ' + xhr.readyState + ')');
                        return xhr;
                    }

                    if (xhr.status === x3dom.nodeTypes.Inline.AwaitTranscoding) {
                        if (that.count < that.numRetries)
                        {
                            that.count++;
                            var refreshTime = +xhr.getResponseHeader("Refresh") || 5;
                            x3dom.debug.logInfo('XHR status: ' + xhr.status + ' - Await Transcoding (' + that.count + '/' + that.numRetries + '): ' + 
                                                'Next request in ' + refreshTime + ' seconds');
                      
                            window.setTimeout(function() {
                                that._nameSpace.doc.downloadCount -= 1;
                                that.loadExternalXml();
                            }, refreshTime * 1000);
                            return xhr;
                        }
                        else
                        {
                            x3dom.debug.logError('XHR status: ' + xhr.status + ' - Await Transcoding (' + that.count + '/' + that.numRetries + '): ' + 
                                                 'No Retries left');
                            that._nameSpace.doc.downloadCount -= 1;
                            that.count = 0;
                            return xhr;
                        }
                    }
                    else if ((xhr.status !== 200) && (xhr.status !== 0)) {
                        that.fireEvents("error");
                        x3dom.debug.logError('XHR status: ' + xhr.status + ' - XMLHttpRequest requires web server running!');

                        that._nameSpace.doc.downloadCount -= 1;
                        that.count = 0;
                        return xhr;
                    }
                    else if ((xhr.status == 200) || (xhr.status == 0)) {
                        that.count = 0;
                    }

                    x3dom.debug.logInfo('Inline: downloading '+that._vf.url[0]+' done.');

					that.loadExternalSuccess(xhr);
					that._nameSpace.doc.downloadCount -= 1;
					that.fireEvents("load");
                    return xhr;
                };

                if (this._vf.url.length && this._vf.url[0].length)
                {
                    var xhrURI = this._nameSpace.getURL(this._vf.url[0]);
                    xhr.open('GET', xhrURI, true);
                    this._nameSpace.doc.downloadCount += 1;
                    try {
                        xhr.send(null);
                    }
                    catch(ex) {
                        this.fireEvents("error");
                        x3dom.debug.logError(this._vf.url[0] + ": " + ex);
                    }
                }
            },


/*
 *	Use MACRO expansion of embedded code to achieve results
 *
 *	Grab namespaceName and params from fields
 *	Supplied function does all work of inserting nodes into DOM/X3D-scenegraph
 *
 *	loadExternalXml probably should be method in Networking class and used by Networking nodes
 *	InsertSceneFragment should be a x3dom library method
 */
			loadExternalSuccess: function (xhr) 
			{
				x3dom.debug.logInfo ('loadExternal success routine');
				x3dom.debug.logInfo ('NamespaceName = ' + this._vf.nameSpaceName.toString());
				x3dom.debug.logInfo ('Parameters = ' + this._vf.params.toString());
				var params = [];
				var pieces = [];
				for (var ii=0; ii<this._vf.params.length; ii++) {
					pieces = this._vf.params[ii].toString().split('=');
					params[pieces[0]] = pieces[1];
				}
				x3dom.debug.logInfo ('Creating NV arry complete');
				var inlMarkup = ParseMacro (xhr.responseText, params);
				
//  --> Get XML markup inside the first Scene element
//				var inlScene = this.getInlineSceneNodes (inlMarkup);
				var inlScene = MarkupToXml (inlMarkup);

// --> If XML (meaning X3D code) parses, then append the code to this node.
				if (inlScene !== undefined && inlScene !== null) {
					this._xmlNode.appendChild (inlScene);
				} else {
					that.fireEvents("error");
                };
				inlScene = null;

			},

// --> Parse X3D code from string to XML
			getInlineScene(code) {
				var parser = new DOMParser();
				var xml = parser.parseFromString (code, "text/xml");
				if (xml !== undefined && xml !== null) {
					var inlScene = xml.getElementsByTagName('Scene')[0] || xml.getElementsByTagName('scene')[0];
					var subthis = new Array();
					subthis.serializer = new XMLSerializer();
					subthis.code = '';
					Array.forEach ( inlScene.childNodes, function (childDomNode)
						{
							if (childDomNode instanceof Element) {
								this['code'] += this.serializer.serializeToString(childDomNode);
							}
						}, subthis);
					return subthis.code;
				}
				return '';
			},
			
			
            loadMacro: function ()
            {
				this.loadExternalXml ();
                var inlScene = null, newScene = null, nameSpace = null, xml = null;
				return;
            }
        }
    )
);

x3dom.nodeTypes.Macro.AwaitTranscoding = 202;      // Parameterizable retry state for Transcoder
x3dom.nodeTypes.Macro.MaximumRetries = 15;         // Parameterizable maximum number of retries

/*
 *	Macro expansion code. This eventually will become a new "node"
 */
function ParseMacro (code, params) {
	var xml = null, inlScene = null;
	var fields = Object.getOwnPropertyNames(params);
	var mcode = code;
	var re;
	for (var ii=0; ii<fields.length; ii++) {
		re = new RegExp ("\%" + fields[ii] + "\%", "g");
		mcode = mcode.replace(re, params[fields[ii]]);
	}
	return mcode;
}
function MarkupToXml (markup) {
	var parser = new DOMParser();
//	var xml = parser.parseFromString ('<Scene>' + markup + '</Scene>', "text/xml");
	var xml = parser.parseFromString (markup, "text/xml");
	if (xml !== undefined && xml !== null) {
		inlScene = xml.getElementsByTagName('Scene')[0] || xml.getElementsByTagName('scene')[0];

	} else {
		if (xml && xml.localName)
			x3dom.debug.logError('No Scene in ' + xml.localName);
		else
			x3dom.debug.logError('No Scene in resource');
	}

	return inlScene;
}
